In [ ]:
# Load CSV file from my google drive 

from google.colab import drive
drive.mount('/content/drive')

In [ ]:
# import pandas library
# read csv file

import pandas as pd

path = "/content/drive/MyDrive/Colab Notebooks/iris.csv"

iris_df = pd.read_csv(path)


In [ ]:
iris_df

5.1  3.5  1.4  0.2     Iris-setosa
0    4.9  3.0  1.4  0.2     Iris-setosa
1    4.7  3.2  1.3  0.2     Iris-setosa
2    4.6  3.1  1.5  0.2     Iris-setosa
3    5.0  3.6  NaN  0.2     Iris-setosa
4    5.4  3.9  1.7  0.4     Iris-setosa
..   ...  ...  ...  ...             ...
144  6.7  3.0  5.2  2.3  Iris-virginica
145  6.3  2.5  5.0  1.9  Iris-virginica
146  6.5  3.0  5.2  NaN  Iris-virginica
147  6.2  3.4  5.4  2.3  Iris-virginica
148  5.9  3.0  5.1  1.8  Iris-virginica

[149 rows x 5 columns]

In [ ]:
iris_df.loc[len(iris_df.index)] = [5.1,	3.5,	1.4,	0.2,	'Iris-setosa']

display(iris_df)

5.1  3.5  1.4  0.2     Iris-setosa
0    4.9  3.0  1.4  0.2     Iris-setosa
1    4.7  3.2  1.3  0.2     Iris-setosa
2    4.6  3.1  1.5  0.2     Iris-setosa
3    5.0  3.6  NaN  0.2     Iris-setosa
4    5.4  3.9  1.7  0.4     Iris-setosa
..   ...  ...  ...  ...             ...
145  6.3  2.5  5.0  1.9  Iris-virginica
146  6.5  3.0  5.2  NaN  Iris-virginica
147  6.2  3.4  5.4  2.3  Iris-virginica
148  5.9  3.0  5.1  1.8  Iris-virginica
149  5.1  3.5  1.4  0.2     Iris-setosa

[150 rows x 5 columns]

In [ ]:
iris_df = iris_df.rename(columns={"5.1": "sepal length", "3.5": "sepal width", 
                                  "1.4": "petal length", "0.2": "petal width", 
                                  "Iris-setosa": "class"})

In [ ]:
display(iris_df)

sepal length  sepal width  petal length  petal width           class
0             4.9          3.0           1.4          0.2     Iris-setosa
1             4.7          3.2           1.3          0.2     Iris-setosa
2             4.6          3.1           1.5          0.2     Iris-setosa
3             5.0          3.6           NaN          0.2     Iris-setosa
4             5.4          3.9           1.7          0.4     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.3          2.5           5.0          1.9  Iris-virginica
146           6.5          3.0           5.2          NaN  Iris-virginica
147           6.2          3.4           5.4          2.3  Iris-virginica
148           5.9          3.0           5.1          1.8  Iris-virginica
149           5.1          3.5           1.4          0.2     Iris-setosa

[150 rows x 5 columns]

In [ ]:
iris_df = iris_df.dropna(axis=1)

In [ ]:
for i in iris_df:
  print(i, iris_df.isna().values.any())

sepal length False
sepal width False
petal length False
petal width False
class False


In [ ]:
display(iris_df)

sepal length  sepal width  petal length  petal width           class
0             4.9          3.0           1.4          0.2     Iris-setosa
1             4.7          3.2           1.3          0.2     Iris-setosa
2             4.6          3.1           1.5          0.2     Iris-setosa
4             5.4          3.9           1.7          0.4     Iris-setosa
5             4.6          3.4           1.4          0.3     Iris-setosa
..            ...          ...           ...          ...             ...
144           6.7          3.0           5.2          2.3  Iris-virginica
145           6.3          2.5           5.0          1.9  Iris-virginica
147           6.2          3.4           5.4          2.3  Iris-virginica
148           5.9          3.0           5.1          1.8  Iris-virginica
149           5.1          3.5           1.4          0.2     Iris-setosa

[146 rows x 5 columns]

## **KNN CLASSIFICATION**

In [ ]:
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, mean_squared_error, mean_absolute_error
from sklearn import preprocessing
from sklearn.model_selection import train_test_split

In [ ]:
train, test = train_test_split(iris_df, test_size = 0.3)

In [ ]:
KNN = KNeighborsClassifier(n_neighbors=5).fit(train[["sepal length", "sepal width",
                                                     "petal length","petal width"]]
                                                    , train["class"])

In [ ]:
flower_classification_test = KNN.predict(test[["sepal length", "sepal width",
                                               "petal length","petal width"]])

flower_classification_train = KNN.predict(train[["sepal length", "sepal width",
                                                 "petal length","petal width"]])

accuracy_testdata = accuracy_score(test["class"], flower_classification_test)

accuracy_traindata = accuracy_score(train["class"], flower_classification_train)

print(" Accuracy Test Data : %.2f" % accuracy_testdata)

print(" Accuracy Train Data : %.2f" % accuracy_traindata)


 Accuracy Test Data : 0.98
 Accuracy Train Data : 0.96


## **Logistic Regression**

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, f1_score

In [ ]:
LGR = LogisticRegression(C = 0.03, solver = 'liblinear').fit(train[["sepal length", "sepal width",
                                                                    "petal length","petal width"]], train["class"])

In [ ]:
flower_classification_test2 = LGR.predict(test[["sepal length", "sepal width",
                                                "petal length","petal width"]])

flower_classification_train2 = LGR.predict(train[["sepal length", "sepal width",
                                                  "petal length","petal width"]])

accuracy_testdata2 = accuracy_score(test["class"], flower_classification_test2)

accuracy_traindata2 = accuracy_score(train["class"], flower_classification_train2)

print(" Accuracy Test Data : %.2f" % accuracy_testdata2)

print(" Accuracy Train Data : %.2f" % accuracy_traindata2)

 Accuracy Test Data : 0.93
 Accuracy Train Data : 0.90


## **SVM**

In [ ]:
from sklearn import svm

In [ ]:
SVM = svm.SVC().fit(train[["sepal length", "sepal width",
                           "petal length","petal width"]], train["class"])

In [ ]:
flower_classification_test3 = SVM.predict(test[["sepal length", "sepal width",
                                                "petal length","petal width"]])

flower_classification_train3 = SVM.predict(train[["sepal length", "sepal width",
                                                  "petal length","petal width"]])

accuracy_testdata3 = accuracy_score(test["class"], flower_classification_test3)

accuracy_traindata3 = accuracy_score(train["class"], flower_classification_train3)

print(" Accuracy Test Data : %.2f" % accuracy_testdata3)

print(" Accuracy Train Data : %.2f" % accuracy_traindata3)

 Accuracy Test Data : 0.98
 Accuracy Train Data : 0.93


Overall, all the machine learning models are not underfitting or overfitting.

The KNN model performs the best in terms of accuracy on both test and training data with an accuracy score of 0.98 and 0.96 respectively. 

The logistic regression model performs the worst out of all 3 models with an accuracy score of 0.93 and 0.90 on test and training data respectively.

The SVM model performs just as well as KNN on the test data, but performs worse on the training data with an accuracy score of 0.93 compared to the KNN's score of 0.96.